edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -0,0 +1,423 @@
+
+            this.forwardingLoop = Option.Some(this.StartForwardingLoop());
+
+            // if ConnectAsync is supposed to manage starting it with broker down,
+            // put a loop here to keep trying - see 'TriggerReconnect' below
+            var isConnected = await TryConnectAsync(client, this.components.Consumers, this.systemComponentIdProvider.EdgeHubBridgeId);
+
+            if (!isConnected)
+            {
+                client.MqttMsgPublished -= this.ConfirmPublished;
+                client.MqttMsgPublishReceived -= this.ForwardPublish;
+
+                await this.StopForwardingLoopAsync();
+
+                lock (this.guard)
+                {
+                    this.mqttClient = Option.None<MqttClient>();
+                }
+
+                throw new Exception("Failed to start mqtt-bridge connector");
+            }
+
+            client.ConnectionClosed += this.TriggerReconnect;
+
+            Events.Started();
+        }
+
+        public async Task DisconnectAsync()
+        {
+            Events.Closing();
+
+            Option<MqttClient> clientToStop;
+
+            lock (this.guard)
+            {
+                clientToStop = this.mqttClient;
+                this.mqttClient = Option.None<MqttClient>();
+            }
+
+            try
+            {
+                await clientToStop.ForEachAsync(
+                        async c =>
+                        {
+                            c.MqttMsgPublished -= this.ConfirmPublished;
+                            c.MqttMsgPublishReceived -= this.ForwardPublish;
+                            c.ConnectionClosed -= this.TriggerReconnect;
+
+                            c.Disconnect();
+
+                            await this.StopForwardingLoopAsync();
+
+                            Events.Closed();
+                        },
+                        () =>
+                        {
+                            Events.ClosedWhenNotRunning();
+                            throw new InvalidOperationException("Cannot stop mqtt-bridge connector when not running");
+                        });
+            }
+            finally
+            {
+                foreach (var tcs in this.pendingAcks.Values)
+                {
+                    tcs.SetCanceled();
+                }
+
+                this.pendingAcks.Clear();
+            }
+        }
+
+        public async Task<bool> SendAsync(string topic, byte[] payload)
+        {
+            var client = this.mqttClient.Expect(() => new Exception("No mqtt-bridge connector instance found to send messages."));
+
+            var added = default(bool);
+            var tcs = new TaskCompletionSource<bool>();
+
+            // need the lock, otherwise it can happen the the ACK comes back sooner as the id is
+            // put into the dictionary next line, causeing the ACK being unknown.
+            lock (this.guard)
+            {
+                var messageId = client.Publish(topic, payload, 1, false);
```

Comment:
This is sync?

Reply:
Yes, the library has sync interface, but in the background it just puts it into a queue, so it is pretty fast